src/chunks.rs

```diff
@@ -0,0 +1,746 @@
+//!         chunk.commit_all();
+//!         Ok(())
+//!     } else {
+//!         Err(slice)
+//!     }
+//! }
+//! ```
+//!
+//! Copy as many items as possible from a given slice, returning the remainder of the slice
+//! (which will be empty if there was space for all items):
+//!
+//! ```
+//! use rtrb::{Producer, chunks::ChunkError::TooFewSlots};
+//!
+//! fn push_partial_slice<'a, T>(queue: &mut Producer<T>, slice: &'a [T]) -> &'a [T]
+//! where
+//!     T: Copy + Default,
+//! {
+//!     let mut chunk = match queue.write_chunk(slice.len()) {
+//!         Ok(chunk) => chunk,
+//!         // This is an optional optimization if the queue tends to be full:
+//!         Err(TooFewSlots(0)) => return slice,
+//!         // Remaining slots are returned, this will always succeed:
+//!         Err(TooFewSlots(n)) => queue.write_chunk(n).unwrap(),
+//!     };
+//!     let end = chunk.len();
+//!     let (first, second) = chunk.as_mut_slices();
+//!     let mid = first.len();
+//!     first.copy_from_slice(&slice[..mid]);
+//!     second.copy_from_slice(&slice[mid..end]);
+//!     chunk.commit_all();
+//!     &slice[end..]
+//! }
+//! ```
+//!
+//! Write as many slots as possible, given an iterator
+//! (and return the number of written slots):
+//!
+//! ```
+//! use rtrb::{Producer, chunks::ChunkError::TooFewSlots};
+//!
+//! fn push_from_iter<T, I>(queue: &mut Producer<T>, iter: &mut I) -> usize
+//! where
+//!     T: Default,
+//!     I: Iterator<Item = T>,
+//! {
+//!     let n = match iter.size_hint() {
+//!         (_, None) => queue.slots(),
+//!         (_, Some(n)) => n,
+//!     };
+//!     let mut chunk = match queue.write_chunk(n) {
+//!         Ok(chunk) => chunk,
+//!         // As above, this is an optional optimization:
+//!         Err(TooFewSlots(0)) => return 0,
+//!         // As above, this will always succeed:
+//!         Err(TooFewSlots(n)) => queue.write_chunk(n).unwrap(),
+//!     };
+//!     for (source, target) in iter.zip(&mut chunk) {
+//!         *target = source;
+//!     }
+//!     chunk.commit_iterated()
+//! }
+//! ```
+
+use core::fmt;
+use core::mem::MaybeUninit;
+use core::sync::atomic::Ordering;
+
+use crate::{Consumer, CopyToUninit, Producer};
+
+// This is used in the documentation.
+#[allow(unused_imports)]
+use crate::RingBuffer;
+
+impl<T> Producer<T> {
+    /// Returns `n` slots (initially containing their [`Default`] value) for writing.
+    ///
+    /// If not enough slots are available, an error
+    /// (containing the number of available slots) is returned.
+    ///
+    /// The elements can be accessed with [`WriteChunk::as_mut_slices()`] or
+    /// by iterating over (a `&mut` to) the [`WriteChunk`].
+    ///
+    /// The provided slots are *not* automatically made available
+    /// to be read by the [`Consumer`].
+    /// This has to be explicitly done by calling [`WriteChunk::commit()`],
+    /// [`WriteChunk::commit_iterated()`] or [`WriteChunk::commit_all()`].
```

Comment:
What happens if `commit`, `commit_iterated()` or `commit_all` is _not_ called? My guess is that the consumer cannot see the items that have been written and that the items written can be overwritten by calling "push()" (or similar) again, which would cause them to be leaked. Is that correct?

Reply:
Exactly.

I've just added a commit (7bb6e5d012d04f2cf7d4e40cf1c49fdd91fb79e9) that hopefully clarifies that.